document add_docker_metadata fields and fix labels.dedot default (#6573)
elastic/docs-content: reference/fleet/add_docker_metadata-processor.md

For events to be annotated with Docker metadata, the configuration must be valid, and the processor must be able to reach the Docker API.

Each event is annotated with the following fields:

| Field | Description |
| --- | --- |
| `container.id` | Container ID |
| `container.name` | Container name |
| `container.image.name` | Container image name |

> | `container.labels.<key>` | One field per Docker label. By default, dots (`.`) in label keys are replaced with underscores (`_`). Refer to [`labels.dedot`](#labels-dedot-behavior) for more information. |

Reviewer: I'm not sure `labels.dedot` is the correct text for this link. I followed the link expecting to see the documentation for `lablels.dedot`, but landed on a section about "Working with labels".

I'd change it to
```suggestion
| `container.labels.<key>` | One field per Docker label. By default, dots (`.`) in label keys are replaced with underscores (`_`). Refer to [working with labels](#labels-dedot-behavior) section for more information. |
```

Author: You're right, I think I rewrote it but didn't change the link. Thanks for catching that!

### Working with labels [labels-dedot-behavior]

> Docker label keys often contain dots, for example `com.docker.compose.service`. By default (`labels.dedot: true`), the processor replaces dots with underscores when writing labels to the event:

Reviewer 2: ```suggestion
Docker label keys often contain dots, for example `com.docker.compose.service`. By default, the processor replaces dots with underscores when writing labels to the event:
```

Reviewer 3: ++ to that

| Original Docker label | Field name in event |
| --- | --- |
| `com.docker.compose.service` | `container.labels.com_docker_compose_service` |
| `com.docker.swarm.service.name` | `container.labels.com_docker_swarm_service_name` |

To preserve the original key names with dots, set `labels.dedot` to `false`. In that case, the labels are stored as nested objects (for example, `container.labels.com.docker.compose.service`).

::::{note}
Creating a field with a name that is dynamically determined at runtime (for example, using the value of a label as a field name) is not supported by this processor or by `add_fields`. To react to label values, use [conditions-based autodiscover templates](/reference/fleet/conditions-based-autodiscover.md) instead.

> ::::

Reviewer 3: Very good note. I think we should include it in the `add_kubernetes_metadata` processor too.

One caveat of the note is that we are linking to the doc about kubernetes autodiscovery, not docker autodiscovery (but docker autodiscovery doc doesn't exist at `fleet` level, only at individual beats documentation)

Author: I added the note to the kubernetes processor page like you suggested. 

Regarding the link, I changed it to dynamic-input-configuration.md, although you're right and what would really work here would be a docker autodiscovery doc on fleet level. Since we don't have that (yet) I tried to find the next best thing that could be useful to the users in this context.

> The Docker provider exposes the same labels as variables for configuration templating (`${docker.container.labels.com.docker.compose.service}`). The provider preserves dots in the key for variable references, regardless of the `labels.dedot` setting. For more information, refer to [Docker provider](/reference/fleet/docker-provider.md).

Reviewer: Maybe this is my engineer brain who knows the internals of Filebeat talking, take this with a grain of salt or just ignore it ;)

I don't like the 'preserves' verb here.... Variable resolution is not related to the fields defined in the events, but this paragraph makes it sound like otherwise.
```suggestion
The Docker provider exposes the same labels as variables for configuration templating (`${docker.container.labels.com.docker.compose.service}`). The provider always references keys in the dotted format. For more information, refer to [Docker provider](/reference/fleet/docker-provider.md).
```

The Docker provider runs way before any processor, it uses its variables to render an input configuration, once the input is running, any defined processor will enrich/modify the ingested events, that's when the `add_docker_meteadata` processor actually runs.

Author: It's better to be precise! That's one of the reasons why we need the review - for the engineering 🧠
